src/main/java/frc/robot/systems/DriveFSMSystem.java

```diff
@@ -0,0 +1,425 @@
+					/ currentTranslationMag);
+			} else {
+				directionSlewRate = DriveConstants.INSTANTANEOUS_SLEW_RATE;
+				//some high number that means the slewrate is effectively instantaneous
+			}
+
+			double currentTime = WPIUtilJNI.now() * DriveConstants.TIME_CONSTANT;
+			double elapsedTime = currentTime - prevTime;
+			double angleDif = SwerveUtils.angleDifference(inputTranslationDir,
+				currentTranslationDir);
+
+			if (angleDif < DriveConstants.ANGLE_MULTIPLIER_1 * Math.PI) {
+				currentTranslationDir = SwerveUtils.stepTowardsCircular(currentTranslationDir,
+					inputTranslationDir, directionSlewRate * elapsedTime);
+				currentTranslationMag = magLimiter.calculate(inputTranslationMag);
+			} else if (angleDif > DriveConstants.ANGLE_MULTIPLIER_2 * Math.PI) {
+				if (currentTranslationMag > DriveConstants.CURRENT_THRESHOLD) {
+					// some small number to avoid floating-point errors with equality checking
+					// keep currentTranslationDir unchanged
+					currentTranslationMag = magLimiter.calculate(0.0);
+				} else {
+					currentTranslationDir = SwerveUtils.wrapAngle(currentTranslationDir + Math.PI);
+					currentTranslationMag = magLimiter.calculate(inputTranslationMag);
+				}
+			} else {
+				currentTranslationDir = SwerveUtils.stepTowardsCircular(currentTranslationDir,
+					inputTranslationDir, directionSlewRate * elapsedTime);
+				currentTranslationMag = magLimiter.calculate(0.0);
+			}
+			prevTime = currentTime;
+
+			xSpeedCommanded = currentTranslationMag * Math.cos(currentTranslationDir);
+			ySpeedCommanded = currentTranslationMag * Math.sin(currentTranslationDir);
+			currentRotation = rotLimiter.calculate(rot);
+
+		} else {
+			xSpeedCommanded = xSpeed;
+			ySpeedCommanded = ySpeed;
+			currentRotation = rot;
+		}
+
+		// Convert the commanded speeds into the correct units for the drivetrain
+		double xSpeedDelivered = xSpeedCommanded * DriveConstants.MAX_SPEED_METERS_PER_SECOND;
+		double ySpeedDelivered = ySpeedCommanded * DriveConstants.MAX_SPEED_METERS_PER_SECOND;
+		double rotDelivered = currentRotation * DriveConstants.MAX_ANGULAR_SPEED;
+
+		var swerveModuleStates = DriveConstants.DRIVE_KINEMATICS.toSwerveModuleStates(
+			fieldRelative
+				? ChassisSpeeds.fromFieldRelativeSpeeds(xSpeedDelivered, ySpeedDelivered,
+					rotDelivered, Rotation2d.fromDegrees(-gyro.getAngle()))
+				: new ChassisSpeeds(xSpeedDelivered, ySpeedDelivered, rotDelivered));
+		SwerveDriveKinematics.desaturateWheelSpeeds(
+			swerveModuleStates, DriveConstants.MAX_SPEED_METERS_PER_SECOND);
+
+		frontLeft.setDesiredState(swerveModuleStates[0]);
+		frontRight.setDesiredState(swerveModuleStates[1]);
+		rearLeft.setDesiredState(swerveModuleStates[2]);
+		rearRight.setDesiredState(swerveModuleStates[(2 + 1)]);
+	}
+
+	/**
+	 * Balances gyro.
+	 */
+	public void balance() {
+		double power;
+		if (Math.abs(gyro.getRoll()) < 2 && Math.abs(gyro.getRoll()) > (-1 * 2)) {
+			power = 0;
+		} else if (gyro.getRoll() > 0) {
```

Comment:
These last three cases are exactly the same. If val < 0, then a + val is the same as a - abs(value).

Also spelling for BALENCE_SPEED_INVERSE_PROPORTION